src/types/custom.rs

```diff
@@ -1,27 +1,35 @@
 //! Opaque types, used to represent a user-defined [`SimpleType`].
+//!
+//! [`SimpleType`]: super::SimpleType
 use smol_str::SmolStr;
 use std::fmt::{self, Display};
 
-use super::{ClassicType, SimpleType, TypeRow};
+use super::{type_param::TypeArg, ClassicType};
 
 /// An opaque type element. Contains an unique identifier and a reference to its definition.
 //
 // TODO: We could replace the `Box` with an `Arc` to reduce memory usage,
 // but it adds atomic ops and a serialization-deserialization roundtrip
 // would still generate copies.
-#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+#[derive(Debug, PartialEq, Eq, Clone, serde::Serialize, serde::Deserialize)]
 pub struct CustomType {
     /// Unique identifier of the opaque type.
-    id: SmolStr,
-    params: Box<TypeRow>,
+    /// Same as the corresponding [`TypeDef`]
+    ///
+    /// [`TypeDef`]: crate::resource::TypeDef
+    pub id: SmolStr,
+    /// Arguments that fit the [`TypeParam`]s declared by the typedef
+    ///
+    /// [`TypeParam`]: super::type_param::TypeParam
+    pub params: Vec<TypeArg>,
```

Comment:
We don't have any current uses, but I'd be surprised if custom code handling these types (giving these "external"/"opaque" declarations actual meaning) didn't need to see the params! We don't really know what that looks like (it's probably within some binary code that performs a custom OpDef *on* those types), so maybe there's another way. I could have a public accessor method instead, but then I feel like I should have a public mutator too, and...accessor/mutator is a bit of a YAGNI (preserving the interface if we change the implementation not to store the actual params, or to store some representation thereof), right?

Reply:
Yeah, I don't like exposing complex internals for API stability reasons (that `Vec` may be a target for perf experiments in a distant future).
Though having the `Serialize` derive kind of exposes everything anyways :P

Reply:
Yeah, that is definitely a downside of this serialization scheme...

I think it might be better to review what's public and isn't when we do a ("stable") release. A quick regex search suggests we have quite a lot of public fields all over the place, which is to say, trying to keep such APIs consistent (in the absence of making such a release), tends to be somewhat patchy. But I've fixed these two. There are a bunch more in resource.rs too tho...